ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/subject/confirmations/BearerSubjectConfirmationVerification.kt

```diff
@@ -0,0 +1,125 @@
+                .flatMap { it.children(SUBJECT) }
+                .filter { it.children(SUBJECT_CONFIRMATION).isNotEmpty() }
+                .flatMap { it.children(SUBJECT_CONFIRMATION) }
+                .filter { it.attributeText("Method") == BEARER }
+                .map { it to it.parentNode.parentNode }
+                .unzip()
+
+        if (bearerAssertions.isEmpty())
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_g,
+                message = "No bearer SubjectConfirmation elements were found.",
+                node = samlResponseDom)
+
+        if (bearerSubjectConfirmations
+                .filter { it.children(SUBJECT_CONFIRMATION_DATA).isNotEmpty() }
+                .flatMap { it.children(SUBJECT_CONFIRMATION_DATA) }
+                .none {
+                    it.attributeText("Recipient") ==
+                        TestCommon.getServiceUrl(HTTP_POST, samlResponseDom) &&
+                        it.attributeNode("NotOnOrAfter") != null &&
+                        it.attributeNode("NotBefore") == null &&
+                        it.attributeText("InResponseTo") == REQUEST_ID
+                }) {
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_h,
+                message = "There were no bearer SubjectConfirmation elements that matched " +
+                    "the criteria below.",
+                node = samlResponseDom)
+        }
+        return bearerAssertions
+    }
+
+    private fun verifyBearerAssertions(bearerAssertions: List<Node>) {
+        if (bearerAssertions.all { it.children(AUTHN_STATEMENT).isEmpty() })
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_i,
+                message = "A Bearer Assertion with an AuthnStatement was not found.",
+                node = samlResponseDom)
+
+        if (idpMetadata.descriptor?.singleLogoutServices?.isEmpty() == true) return
+
+        if (bearerAssertions
+                .flatMap { it.children(AUTHN_STATEMENT) }
+                .any { it.attributeNode("SessionIndex") == null })
+            throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_j,
+                message = "Single Logout support found in IdP metadata, but no " +
+                    "SessionIndex was found for AuthnStatement.",
+                node = samlResponseDom)
+    }
+
+    private fun verifyAudienceRestriction(bearerAssertions: List<Node>) {
+        bearerAssertions.forEach { bearerAssertion ->
+            if (bearerAssertion.children("Conditions")
+                    .map { extractAudienceRestriction(it) }
+                    .filter { it.children(AUDIENCE).isNotEmpty() }
+                    .flatMap { it.children(AUDIENCE) }
+                    .none { it.textContent == currentSPIssuer }) {
+
+                throw SAMLComplianceException.create(SAMLProfiles_4_1_4_2_k,
+                    message = "A bearer assertion which doesn't contains an <Audience> with the " +
```

Comment:
✏️ This message isn't clear. It sounds like it's putting the condition above into English more than describing the incorrect state it found.